Merge 2.12.x to 2.13.x [ci: last-only]

retronym opened the pull request:
```
% export MB=$(git merge-base origin/2.12.x origin/2.13.x)

% git log $MB..origin/2.12.x --graph --decorate --oneline  | tail -r
* 91bf0c78fc (retronym/backport/7671-java-ident, origin/pr/7738) [backport] Use Java rules for member lookup in .java sources
* 9e6ca45c2f Merge pull request #7738 from retronym/backport/7671-java-ident
|/
| * 65857146ca (origin/pr/7774, SethTisue/partest-1.1.9) partest 1.1.9 (was 1.1.7)
|/|
| |/
|\ \
* |   a2aa480be5 Merge pull request #7774 from SethTisue/partest-1.1.9
|/ /
| * | b37c0a42b3 (origin/pr/7786, diesalbla/glb_merge_for_comprehension_map) Small optimisation in glbNorm function.
|/| |
| |/ /
|\ \ \
* | |   a9337c2aa8 Merge pull request #7786 from diesalbla/glb_merge_for_comprehension_map
| |/ /
| * | | a996dc49eb misc minor readme tweaks
| * | | 5ad9e03fde remove inactive maintainers
| * | | 1bb9b7482c readme: reorder maintainers, add Stefan and Viktor
| * | | 6ff01aec2b (origin/pr/7778, SethTisue/maintainers-list) Add sjrd to the maintainer list, for interactions with Scala.js.
|/| | |
| |/ / /
|\ \ \ \
* | | |   9197882eaf Merge pull request #7778 from SethTisue/maintainers-list
| |/ / /
| * | | | e22e12da28 (retronym/topic/unblock-scala-meta, origin/pr/7799, topic/unblock-scala-meta) Restore API in typer used by scala-meta
|\ \ \ \
* | | |   ac535fdbbe Merge pull request #7799 from retronym/topic/unblock-scala-meta
| |/ / /
| * | | | ecfa631545 (origin/pr/7791, diesalbla/glbNorm_performance_tweaks) Small performance tweak to glbNorm.
|\ \ \ \
* | | |   5915f41f5f Merge pull request #7791 from diesalbla/glbNorm_performance_tweaks
| * | | 9a04c4d9b7 (origin/pr/7797, diesalbla/complexity_map_into_fold) Complexity: fold map into the sum.
|\ \ \
* | |   b63ccf2296 Merge pull request #7797 from diesalbla/complexity_map_into_fold
| |/ /
| * | | 6a8177b4e1 [backport] Upgrade to ASM 7
| * | | 1220d3c915 (origin/pr/7780, SethTisue/asm7) [backport] Nicer branch-sensitive nullness
|\ \ \
* | |   ce9a857920 (retronym/2.12.x, topic/2.12.x-backports) Merge pull request #7780 from SethTisue/asm7
| * | a1539c6934 (origin/pr/7788, diesalbla/backports_upperBound_lowerBound) Back-ports changes to add upperBound, lowerBound.
|/| |
| |_|/
|\ \ \
* | |   f83d073843 Merge pull request #7788 from diesalbla/backports_upperBound_lowerBound
| * | cfbb2a2d16 [backport] Resource management for macro/plugin classloaders, classpath JARs
| * | 8efcb4236b unpickler cycle avoidance
| * | a4d9599f1d Utility to abstract over JAR and directory I/O
| * | 69d3c00027 Utility to extract Pickles from Scala compiled classes
| * | b066d7e640 Compute pickles for .java sources
| * | 22eae50c8f .sig file support
| * | 27a976c084 Experimental support for outline typing
| * | c6e290ad4a Driver to build a set of compile @argsfile-s in pipeline.
| * | 8e58ea0bc6 Valid URL for VirtualDirectory classpath
| * | 141a72f6fa Add missing copyright headers
| * | 7707a763fc [nomerge] Restore findMacroClassloader into Analyzer for the 2.12.x series
| * | 0b1974c8e7 Refactor PipelineMain
| * | d9b98b8d74 Whitelist some binary changes to internals of scala-reflect
| * | 2c6a4b22e4 Remove outline typechecking for now
| * | 3e500b21a5 Use file size as part of the cache invalidation
| | |/
| |\ \
| * \   5747be5840 (retronym/topic/pipeline-trace-macro-omnibus-rebase, origin/pr/7712, topic/pipeline-trace-macro-omnibus-rebase) Merge remote-tracking branch 'origin/2.12.x' into topic/pipeline-trace-macro-omnibus-rebase
|\ \
* |   2961eda885 Merge pull request #7712 from retronym/topic/pipeline-trace-macro-omnibus-rebase
| |/
| * | ea0e5c8daf (origin/pr/7810, diesalbla/remove_bothnames) Remove the bothNames method from the Name class.
|\ \
* |   2a395754ce Merge pull request #7810 from diesalbla/remove_bothnames
| * 693f3a724b (origin/pr/7808, diesalbla/tryTypedApply_flatmap_exists) Typers: merge treesInResult with errorInResult
|\
*   29c6d99f28 (diesalbla/2.12.x, 2.12.x) Merge pull request #7808 from diesalbla/tryTypedApply_flatmap_exists
|/
| * 3e0ab870ce (retronym/topic/cache-policy-fix, origin/pr/7813, topic/cache-policy-fix) Honour CachePolicy.None in classloader/classpath caching
|\
*   88ed07f080 (origin/2.12.x) Merge pull request #7813 from retronym/topic/cache-policy-fix

% git merge -sours a2aa480be5
Merge made by the 'ours' strategy.

% git merge b63ccf2296
Auto-merging src/reflect/scala/reflect/internal/tpe/GlbLubs.scala
Auto-merging src/compiler/scala/tools/nsc/typechecker/Typers.scala
Auto-merging src/compiler/scala/tools/nsc/typechecker/Implicits.scala
CONFLICT (content): Merge conflict in src/compiler/scala/tools/nsc/typechecker/Implicits.scala
Auto-merging README.md
Automatic merge failed; fix conflicts and then commit the result.

% git status
On branch merge/2.12.x-to-2.12.x-20180305
All conflicts fixed but you are still merging.

Changes to be committed:
    modified:   README.md
    modified:   src/compiler/scala/tools/nsc/typechecker/Implicits.scala
    modified:   src/compiler/scala/tools/nsc/typechecker/Typers.scala
    modified:   src/reflect/scala/reflect/internal/tpe/GlbLubs.scala


% git add .

% git commit
[merge/2.12.x-to-2.12.x-20180305 98f0002726] Merge commit 'b63ccf2296' into merge/2.12.x-to-2.12.x-20180305

% git merge -sours f83d073843
Merge made by the 'ours' strategy.

% git merge -sours cfbb2a2d16
Merge made by the 'ours' strategy.

% git merge 141a72f6fa
Auto-merging src/reflect/scala/reflect/internal/pickling/UnPickler.scala
Auto-merging src/reflect/scala/reflect/internal/Definitions.scala
Auto-merging src/compiler/scala/tools/nsc/typechecker/Typers.scala
Auto-merging src/compiler/scala/tools/nsc/typechecker/Analyzer.scala
Auto-merging src/compiler/scala/tools/nsc/symtab/classfile/Pickler.scala
Auto-merging src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala
CONFLICT (content): Merge conflict in src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala
Auto-merging src/compiler/scala/tools/nsc/symtab/classfile/AbstractFileReader.scala
Auto-merging src/compiler/scala/tools/nsc/settings/ScalaSettings.scala
CONFLICT (content): Merge conflict in src/compiler/scala/tools/nsc/settings/ScalaSettings.scala
Auto-merging src/compiler/scala/tools/nsc/classpath/VirtualDirectoryClassPath.scala
Auto-merging src/compiler/scala/tools/nsc/Global.scala
Auto-merging src/compiler/scala/tools/nsc/CompilationUnits.scala
Automatic merge failed; fix conflicts and then commit the result.

% git status
gOn branch merge/2.12.x-to-2.12.x-20180305
All conflicts fixed but you are still merging.

Changes to be committed:
    modified:   src/compiler/scala/tools/nsc/CompilationUnits.scala
    modified:   src/compiler/scala/tools/nsc/Global.scala
    new file:   src/compiler/scala/tools/nsc/PickleExtractor.scala
    new file:   src/compiler/scala/tools/nsc/PipelineMain.scala
    modified:   src/compiler/scala/tools/nsc/classpath/FileUtils.scala
    modified:   src/compiler/scala/tools/nsc/classpath/VirtualDirectoryClassPath.scala
    modified:   src/compiler/scala/tools/nsc/settings/ScalaSettings.scala
    modified:   src/compiler/scala/tools/nsc/symtab/classfile/AbstractFileReader.scala
    modified:   src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala
    modified:   src/compiler/scala/tools/nsc/symtab/classfile/Pickler.scala
    modified:   src/compiler/scala/tools/nsc/typechecker/Analyzer.scala
    modified:   src/compiler/scala/tools/nsc/typechecker/Typers.scala
    modified:   src/reflect/scala/reflect/internal/Definitions.scala
    modified:   src/reflect/scala/reflect/internal/pickling/UnPickler.scala
    new file:   src/reflect/scala/reflect/io/RootPath.scala


% git add .

% git commit
[merge/2.12.x-to-2.12.x-20180305 aa3ffb71c3] Merge commit '141a72f6fa' into merge/2.12.x-to-2.12.x-20180305

% git merge -sours 7707a763fc
Merge made by the 'ours' strategy.

% git merge 88ed07f080
Auto-merging src/repl/scala/tools/nsc/interpreter/MemberHandlers.scala
Auto-merging src/reflect/scala/reflect/internal/Names.scala
CONFLICT (modify/delete): src/reflect/mima-filters/2.12.0.forwards.excludes deleted in HEAD and modified in 88ed07f080. Version 88ed07f080 of src/reflect/mima-filters/2.12.0.forwards.excludes left in tree.
CONFLICT (modify/delete): src/reflect/mima-filters/2.12.0.backwards.excludes deleted in HEAD and modified in 88ed07f080. Version 88ed07f080 of src/reflect/mima-filters/2.12.0.backwards.excludes left in tree.
Auto-merging src/compiler/scala/tools/nsc/typechecker/Typers.scala
Auto-merging src/compiler/scala/tools/nsc/typechecker/Namers.scala
Auto-merging src/compiler/scala/tools/nsc/typechecker/Analyzer.scala
Auto-merging src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala
CONFLICT (content): Merge conflict in src/compiler/scala/tools/nsc/classpath/ZipAndJarFileLookupFactory.scala
Automatic merge failed; fix conflicts and then commit the result.

% git commit
[merge/2.12.x-to-2.12.x-20180305 14ad07234a] Merge commit '88ed07f080' into merge/2.12.x-to-2.12.x-20180305


```

--------------

```
$ git log 88ed07f080..origin/2.12.x --oneline --graph --decorate | tail -r
* 57277be9a7 (origin/pr/7816, diesalbla/fix_add_deprecation_bothnames) Restore deprecated bothnames method.
* 9862ab7a15 Merge pull request #7816 from diesalbla/fix_add_deprecation_bothnames
|/
| * d62f26c982 (retronym/topic/closeable-registry, origin/pr/7825, topic/closeable-registry) Avoid NPE during global initialization under -verbose/-Ylogcp
|\
*   f0936e05ab Merge pull request #7825 from retronym/topic/closeable-registry
|/
| * ca8c69da8e (origin/pr/7822, SethTisue/pathresolver-backcompat) restore compat for PathResolver, ClassPathFactory constructors
|/|
| |/
|\ \
* |   7ae699585e Merge pull request #7822 from SethTisue/pathresolver-backcompat
|  /
| * | 1494b64d74 Avoid the parentSymbols method: less allocations.
| * | a789144411 (origin/pr/7823, diesalbla/parentsymbols_remove) Add parentSymbolsIterator method. Use it instead of the info.
|\ \
* |   25c7d20ce5 Merge pull request #7823 from diesalbla/parentsymbols_remove
|  /
| * | 9fcb2f34fc (origin/pr/7824, diesalbla/typers_avoid_flatten) Avoid using List.flatten
|\ \
* |   2b9aa1c264 (retronym/2.12.x) Merge pull request #7824 from diesalbla/typers_avoid_flatten
| * 8763166e8f (retronym/topic/picklewriter, origin/pr/7828, topic/picklewriter) Avoid trailing zero bytes in .sig files written in PipelineMain
|/|
| |/
|\ \
* |   8e71ae3348 Merge pull request #7828 from retronym/topic/picklewriter
|/ /
| * | fcb8ff7a29 (origin/pr/7837, diesalbla/copy_method_completer_replace_map2_with_foreach2) copyMethodCompleter: Replace map2 with foreach2
|/| |
| |/ /
|\ \ \
* | |   6b1fdbb7de Merge pull request #7837 from diesalbla/copy_method_completer_replace_map2_with_foreach2
| |/ /
| * | | 524c672ef2 (origin/pr/7838, diesalbla/specialize_types_avoid_zip) Avoid using List.zip in the SpecializeTypes unification
|\ \ \
* | |   56d29c4a80 Merge pull request #7838 from diesalbla/specialize_types_avoid_zip
| |/ /
| * | | 57c4888f08 Remove use of List.map in the "varianceInTypes" function.
| * | | 150e0fe350 (retronym/review/7835) Reduce SAM lambda allocation within varianceInTypes
| * | | e6cc02db5d (origin/pr/7835, diesalbla/variances_remove_listmap) Make varianceInType allocation free with a reusable instance
|\ \ \
* | |   2a18ab80a7 Merge pull request #7835 from diesalbla/variances_remove_listmap
| |/ /
| * | | 3cc07a4955 (origin/pr/7827, diesalbla/uncurry_avoid_flatten) Uncurry: avoid using List.flatten
|\ \ \
* | |   e214fe3577 Merge pull request #7827 from diesalbla/uncurry_avoid_flatten
| * | a181111338 (origin/pr/7809, diesalbla/avoid_list_creation_) Avoid the call to BaseTypeSeq.toList in type Unification.
|\ \
* |   efb9faf7f6 Merge pull request #7809 from diesalbla/avoid_list_creation_
|  /
| * | 4e718b3359 (origin/pr/7820, diesalbla/dedup_no_lists_backport) Selectively Backport some changes
|/| |
| |/ /
|\ \ \
* | |   f6930376f5 Merge pull request #7820 from diesalbla/dedup_no_lists_backport
|  / /
| * | | ca3e491114 Avoid using List flatMap withi normalizeImpl
| * | | d27a93af6e (origin/pr/7826, diesalbla/types_normalize_no_flatmap) Use LinkedHashSet instead of ListBuffer: better query times
|\ \ \
* | |   89f8daf77b Merge pull request #7826 from diesalbla/types_normalize_no_flatmap
| * | 33d9f6f303 (origin/pr/7840, diesalbla/typers_avoid_unzipping) Typers: by-pass the unzipped list of pairs.
|\ \
* |   c04a5dd786 Merge pull request #7840 from diesalbla/typers_avoid_unzipping
|/ /
| * | 6e42ccadda (origin/pr/7852, diesalbla/In_map_substInfo_use_foreach) Replace the uses of map for substInfo
|/| |
| |/ /
|\ \ \
* | |   e3c0c10660 Merge pull request #7852 from diesalbla/In_map_substInfo_use_foreach
| |/ /
| * | | 5319a9e1f5 (origin/pr/7855, diesalbla/merge_find_foreach_map) Superaccessors: merge several List method.
|\ \ \
* | |   3e98d20bf2 Merge pull request #7855 from diesalbla/merge_find_foreach_map
| * | b2bee0c095 (origin/pr/7854, diesalbla/foreachParamss) Optimisation: replace mapParamss with a foreachParamss
|\ \
* |   9f5d27d4a9 Merge pull request #7854 from diesalbla/foreachParamss
|/ /
| * | 288d49a068 (origin/pr/7858, diesalbla/Definitions_replace_map_distinct_check_singleton_with_head_tail_forall) Definitions - SymbolSet class: deforest middle list.
|/| |
| |/ /
|\ \ \
* | |   724aac58d7 Merge pull request #7858 from diesalbla/Definitions_replace_map_distinct_check_singleton_with_head_tail_forall
| |/ /
| * | | 9a41c4cd6f (origin/pr/7868, diesalbla/Checkable_avoid_allocations) Remove List Allocations in the Checkable
|\ \ \
* | |   02d7a464dc Merge pull request #7868 from diesalbla/Checkable_avoid_allocations
| * | 6a58cce5a0 (origin/pr/7865) Optimisations: avoid List allocations in the RefChecks.
|\ \
* |   6f4ee4a51d Merge pull request #7865 from diesalbla/RefChecks_avoid_flatten_maps_filter
|/ /
| * | afe6d85ea9 (origin/pr/7872) [backport] Prefer HTTPS over HTTP for pom.xml data
|\ \
* |   4acff24d35 Merge pull request #7872 from ennru/prefer-https-2.12
| * e9bd65d7f8 Fix completion of explicitly imported type names
| * 55ba885d2a (retronym/topic/completion-new, origin/pr/7833, topic/completion-new) Show constructor def strings TAB,TAB, as we do for methods
|/|
| |/
|\ \
* |   d708101d83 Merge pull request #7833 from retronym/topic/completion-new
| * 0bd07f9f8c (origin/pr/7881, diesalbla/zipped_map_convert_to_map2) Replace "zipped" by combinators without list allocations.
|/|
| |/
|\ \
* |   01cb722a28 Merge pull request #7881 from diesalbla/zipped_map_convert_to_map2
| |/
| * | c5a69fdb50 Don't give unhelpful unemptiness advice.
| * | b87e2c5b3b (origin/pr/7898, hrhino/topic/not-unempty) mailmap me!
|\ \
* |   f967dcd9f0 (diesalbla/2.12.x) Merge pull request #7898 from hrhino/topic/not-unempty
| * 98e59b41c9 (retronym/topic/implicit-shadower-reusable, origin/pr/7890) Recycle the hashset used for implicit shadowing
|/|
| |/
|\ \
* |   eb8280dbd2 (2.12.x) Merge pull request #7890 from retronym/topic/implicit-shadower-reusable
| * cad96e244b Refactor adjustTypeArgs, giving its result record a name
| * 5ba6776250 (retronym/topic/adjusted-args, origin/pr/7900, topic/adjusted-args) Avoid redundant computation of undetParams / refacator to avoid confusing similarly named idents
|\
*   6990ee369a (origin/2.12.x) Merge pull request #7900 from retronym/topic/adjusted-args
```

```
% git merge -sours 9862ab7a15
%  git merge efb9faf7f6
%  git merge -sours f6930376f5
%  git merge 6f4ee4a51d
%  git merge -sours 4acff24d35
%  git merge 6990ee369a
``

lrytz committed:
[backport] Nicer branch-sensitive nullness

lrytz committed:
[backport] Upgrade to ASM 7

retronym committed:
[backport] Use Java rules for member lookup in .java sources

 - Inherited type declarations are in scope in Java code
 - For static innner classes, we need to check in the companion
module of each base classes.
 - Incorporate and accomodate test case from #6053
 - Tests to java code referring to module-class owned classes via companion class prefix

 Backport of scala/scala#7671

retronym committed:
[backport] Resource management for macro/plugin classloaders, classpath JARs

Backports:

  - https://github.com/scala/scala/pull/7366
  - https://github.com/scala/scala/pull/7644

retronym committed:
Utility to abstract over JAR and directory I/O

retronym committed:
Utility to extract Pickles from Scala compiled classes

.sig files, containing the pickle bytes, are output
in place of .class files emitted by Scalac.

Java defined .class files are emitted after stripping
them of code.

retronym committed:
Compute pickles for .java sources

If a new hidden setting is enabled. Build tools that want to implement
build pipelining can use this to feed to downstream classpaths.

retronym committed:
unpickler cycle avoidance

retronym committed:
Experimental support for outline typing

Under this mode, the RHS of defs and vals are only typechecked if
the result type of the definition is inferred and the definition's
signature is forced.

The movitivation is to create a fast path to extract the pickles for
the API for use on the classpath of downstream compiles, which could
include parallel compilation of chunks of the the current set of
source files.

retronym committed:
.sig file support

retronym committed:
Driver to build a set of compile @argsfile-s in pipeline.

retronym committed:
Valid URL for VirtualDirectory classpath

retronym committed:
Add missing copyright headers

retronym committed:
Whitelist some binary changes to internals of scala-reflect

retronym committed:
Merge pull request #7738 from retronym/backport/7671-java-ident

[backport] Use Java rules for member lookup in .java sources

SethTisue committed:
partest 1.1.9 (was 1.1.7)

there aren't any real changes in this version. the context is
to test that we are able to publish Scala modules using sbt 1 now

SethTisue committed:
Merge pull request #7774 from SethTisue/partest-1.1.9

partest 1.1.9 (was 1.1.7)

SethTisue committed:
misc minor readme tweaks

SethTisue committed:
remove inactive maintainers

SethTisue committed:
readme: reorder maintainers, add Stefan and Viktor

diesalbla committed:
Small optimisation in glbNorm function.

The section of code being modified had two inefficiencies:
- It created a `syms` list that was only used in the `map`
  immediatily afterwards.
- It was performing two calls to the `glbThisType.memberInfo` method.

We change this code to fix these inefficiencies: we replace the for
comprehensions with a set of foreach statements, insert on ListBuffer.

diesalbla committed:
Back-ports changes to add upperBound, lowerBound.

Back-ports the changes from https://github.com/scala/scala/pull/7142

We add an `upperBound` and `lowerBound` method to the Type class,
and replace as many calls as we can of `.bounds` with calls to these.

diesalbla committed:
Small performance tweak to glbNorm.

We change some auxiliary methods of glbNorm.

We replace the recursive `refinedToParentsList` function, that
was using nested List.flatMap operations, with a custom function
that uses a mutable ListBuffer and iterates recursively through
the RefinedType elements.

We replace the `refinedToDecls` method, which was building a list of
scopes that was later iterated over, by a method refinedDeclsForeach
that iterates over the elements that would be added to that list.

Signed-off-by: Diego Alonso <diesalbla@gmail.com>

diesalbla committed:
Merge pull request #7786 from diesalbla/glb_merge_for_comprehension_map

Small optimisation in glbNorm function.

The section of code being modified had two inefficiencies:
- It created a `syms` list that was only used in the `map`
  immediatily afterwards.
- It was performing two calls to the `glbThisType.memberInfo` method.

We change this code to fix these inefficiencies: we replace the for
comprehensions with a set of foreach statements, insert on ListBuffer.

diesalbla committed:
Complexity: fold map into the sum.

Merges the use of a `map` function into the `sum` function, which
avoids allocating a list of objects.

sjrd committed:
Add sjrd to the maintainer list, for interactions with Scala.js.

retronym committed:
Restore API in typer used by scala-meta

SethTisue committed:
Merge pull request #7778 from SethTisue/maintainers-list

readme changes

SethTisue committed:
Merge pull request #7799 from retronym/topic/unblock-scala-meta

Restore API in typer used by scala-meta

retronym committed:
[nomerge] Restore findMacroClassloader into Analyzer for the 2.12.x series

retronym committed:
Refactor PipelineMain

  - Scope the build strategies inside object PipelineMain
  - Prefer Future.traverse to Future.sequence

diesalbla committed:
Merge pull request #7791 from diesalbla/glbNorm_performance_tweaks

Small performance tweak to glbNorm.

diesalbla committed:
Merge pull request #7797 from diesalbla/complexity_map_into_fold

Complexity: fold map into the sum.

SethTisue committed:
Merge pull request #7780 from SethTisue/asm7

[backport] upgrade to ASM 7

diesalbla committed:
Typers: merge treesInResult with errorInResult

In the Typers file, in the `tryTypedApply` function, there was a
function treesInResult that was generating a list that contained
each tree and many of its subtrees. The result of this function
was a list that was immediately afterwards put into an exists
function, thus consuming the generated list right away.

To avoid list allocations, we replace that list generation and
traversal with a tree traversal, that carries out the exists function
directly on the nodes of the tree.

diesalbla committed:
Avoid the call to BaseTypeSeq.toList in type Unification.

The method `toList` of the `BaseTypeSeq` class creates a list
from the elements already in an array. In the Types unification,
this method was called to create a list of types that was
immediately filtered out and folded with an `exists`.

To avoid the call to `toList`, we do the following:
- Add a `toIterator` method in the `BaseTypeSeq` class. This is a
  simple counter that goes through the array at once.
- use the iterator.exists method instead.

diesalbla committed:
Remove the bothNames method from the Name class.

The Name class defined a bothNames method which always returned a list
of two elements, the termName and the typeName, both publicly accessible.

To avoid needless List allocations, we remove this method and replace
any use of it by a direct call to the termName and typeName methods.

retronym committed:
Remove outline typechecking for now

We need to typecheck the RHS of type-ascribed definitions if they
contain Super trees that require super-accessors to be added to the
enclosing template. Rather than take that on right now, I'm removing
this feature to focus on the other form of build pipelining.

retronym committed:
Use file size as part of the cache invalidation

retronym committed:
Merge remote-tracking branch 'origin/2.12.x' into topic/pipeline-trace-macro-omnibus-rebase

retronym committed:
Merge pull request #7788 from diesalbla/backports_upperBound_lowerBound

Back-ports changes to add upperBound, lowerBound.

retronym committed:
Merge pull request #7712 from retronym/topic/pipeline-trace-macro-omnibus-rebase

Experimental support for build pipelining [ci: last-only]

diesalbla committed:
Merge pull request #7810 from diesalbla/remove_bothnames

Remove the bothNames method from the Name class.

diesalbla committed:
Merge pull request #7808 from diesalbla/tryTypedApply_flatmap_exists

Typers: merge treesInResult with errorInResult

retronym committed:
Honour CachePolicy.None in classloader/classpath caching

I broke this in a recent refactoring.

retronym committed:
Merge pull request #7813 from retronym/topic/cache-policy-fix

Honour CachePolicy.None in classloader/classpath caching

retronym committed:
Merge commit 'a2aa480be5' into merge/2.12.x-to-2.12.x-20180305

retronym committed:
Merge commit 'b63ccf2296' into merge/2.12.x-to-2.12.x-20180305

retronym committed:
Merge commit 'f83d073843' into merge/2.12.x-to-2.12.x-20180305

retronym committed:
Merge commit 'cfbb2a2d16' into merge/2.12.x-to-2.12.x-20180305

retronym committed:
Merge commit '141a72f6fa' into merge/2.12.x-to-2.12.x-20180305

retronym committed:
Merge commit '7707a763fc' into merge/2.12.x-to-2.12.x-20180305

retronym committed:
Merge commit '88ed07f080' into merge/2.12.x-to-2.12.x-20180305

retronym committed:
Fixup merge errors

diesalbla committed:
Restore deprecated bothnames method.

We restore the method bothNames, with the deprecated annotation, to
prevent compiler plugins inadvertently falling into a binary crash.

SethTisue committed:
Merge pull request #7816 from diesalbla/fix_add_deprecation_bothnames

Restore deprecated bothnames method.

diesalbla committed:
Selectively Backport some changes

We "backport", but with some changes, the improvements to the
`noDuplicates` function already introduced in the 2.13.x branch.
Importantly, this avoids the allocation of two mapped lists.

diesalbla committed:
Avoid the parentSymbols method: less allocations.

The "Type" abstract class defines a method "parentSymbols" that is
implemented by getting the list of parents of the type, and then
performing a "map" to get each parent's type symbol.
This map is allocating a list.

Looking at the usages of this method, we confirmed that most of the
calls to this method were followed by a "fold" on the list, such as an
"contains", or a "find". In some cases, the list is not used at all.

To save allocations, we replace the calls to "parentSymbols" with
the calls to its implementation, "info.parents", and replace the
"contains" methods with "exists(_.typesymbol)".

SethTisue committed:
restore compat for PathResolver, ClassPathFactory constructors

restore source compat and bincompat

small followup to #7712. the community build found that a couple of
projects (mima, classpath-shrinker) were using the old
constructors. since it's easy to do, let's keep both source compat
(with the default arguments) and bincompat (with the extra
constructors, which we can toss for 2.13)

diesalbla committed:
Avoid using List.flatten

The method `List.flatten`, which is applied to a list of lists of
elements, results in a lot of allocations for the concatenation.

In this commit, we remove one use of a call to flatten in the Typers.
A for loop on the flattened list is same as a for loop on the outer
list and one on each inner list. An "exists" on the flattened list is
same as an outer exists of inner exists.

In the same way, when computing the sum of lengths of the lists in a
list of list, we replace the flatten with a custom function for that.

retronym committed:
Avoid NPE during global initialization under -verbose/-Ylogcp

```
$ git clone wheaties/TwoTails
```

```
sbt:root> ; ++2.12.9-bin-88ed07f-SNAPSHOT! ; Test/compile

[error] java.lang.NullPointerException
[error]     at scala.tools.nsc.classpath.FileBasedCache.getOrCreate(ZipAndJarFileLookupFactory.scala:269)
[error]     at scala.tools.nsc.classpath.ZipAndJarFileLookupFactory.create(ZipAndJarFileLookupFactory.scala:44)
[error]     at scala.tools.nsc.classpath.ZipAndJarFileLookupFactory.create$(ZipAndJarFileLookupFactory.scala:37)
[error]     at scala.tools.nsc.classpath.ZipAndJarClassPathFactory$.create(ZipAndJarFileLookupFactory.scala:55)
[error]     at scala.tools.nsc.classpath.ClassPathFactory$.newClassPath(ClassPathFactory.scala:85)
[error]     at scala.tools.nsc.classpath.ClassPathFactory.newClassPath(ClassPathFactory.scala:29)
[error]     at scala.tools.nsc.classpath.ClassPathFactory.$anonfun$classesInPathImpl$3(ClassPathFactory.scala:69)
[error]     at scala.tools.nsc.classpath.ClassPathFactory.$anonfun$classesInPathImpl$1(ClassPathFactory.scala:65)
[error]     at scala.tools.nsc.classpath.ClassPathFactory.classesInPathImpl(ClassPathFactory.scala:64)
[error]     at scala.tools.nsc.classpath.ClassPathFactory.classesInPath(ClassPathFactory.scala:55)
[error]     at scala.tools.util.PathResolver$Calculated$.basis(PathResolver.scala:260)
[error]     at scala.tools.util.PathResolver$Calculated$.containers$lzycompute(PathResolver.scala:272)
[error]     at scala.tools.util.PathResolver$Calculated$.containers(PathResolver.scala:272)
[error]     at scala.tools.util.PathResolver.containers(PathResolver.scala:288)
[error]     at scala.tools.util.PathResolver.computeResult(PathResolver.scala:310)
[error]     at scala.tools.util.PathResolver.result(PathResolver.scala:293)
[error]     at scala.tools.nsc.backend.JavaPlatform.classPath(JavaPlatform.scala:30)
[error]     at scala.tools.nsc.backend.JavaPlatform.classPath$(JavaPlatform.scala:29)
[error]     at scala.tools.nsc.Global$GlobalPlatform.classPath(Global.scala:127)
[error]     at scala.tools.nsc.Global.classPath(Global.scala:138)
[error]     at scala.tools.nsc.Global.<init>(Global.scala:364)
[error]     at xsbt.CallbackGlobal.<init>(CallbackGlobal.scala:21)
[error]     at xsbt.ZincCompiler.<init>(CallbackGlobal.scala:60)
[error]     at xsbt.CachedCompilerCompat.newCompiler(Compat.scala:31)
[error]     at xsbt.CachedCompilerCompat.newCompiler$(Compat.scala:30)
[error]     at xsbt.CachedCompiler0.newCompiler(CompilerInterface.scala:55)
[error]     at xsbt.CachedCompiler0.<init>(CompilerInterface.scala:84)
[error]     at xsbt.CompilerInterface.newCompiler(CompilerInterface.scala:22)
```

SethTisue committed:
Merge pull request #7825 from retronym/topic/closeable-registry

Avoid NPE during global initialization under -verbose/-Ylogcp

diesalbla committed:
Avoid using List flatMap withi normalizeImpl

The normalizeImpl was using a `flatten` method for flattening out the
list of parents of a refined type. This `flatten` method combined a
list map with a list flatMap and a recursive loop, which could give a
lot of list allocations discarded.

We replace the code by a simple `foreach` loop that uses a mutable
list buffer. To further avoid the call to the "distinct" method,
we also check for repeated elements before insertion.

diesalbla committed:
Uncurry: avoid using List.flatten

The `varargForwarderSym` method was using the `List.flatten` method to
reduce two lists of lists of parameters to a single list. Also, in one
of those lists it was using a `list.map`, with extra allocations.
However, these lists were immediately paired in a `foreach` loop, and
thus discarded immediately.

However, since one of the lists is precisely a copy (cloned) from the
other, we can instead replace the `flatten` with a double nested
for-each pairs loop.

retronym committed:
Avoid trailing zero bytes in .sig files written in PipelineMain

diesalbla committed:
Add parentSymbolsIterator method. Use it instead of the info.

diesalbla committed:
Use LinkedHashSet instead of ListBuffer: better query times

SethTisue committed:
Merge pull request #7822 from SethTisue/pathresolver-backcompat

ClassPathFactory and PathResolver constructors: restore source compat and bincompat

retronym committed:
Merge pull request #7823 from diesalbla/parentsymbols_remove

Avoid the parentSymbols method: less allocations.

retronym committed:
Merge pull request #7824 from diesalbla/typers_avoid_flatten

Micro-optimisation: avoid using List.flatten

retronym committed:
Fix completion of explicitly imported type names

The logic under all `allImportedSymbols` differed from `importedSymbol`
by fruitlessly copmaring the import selector (a TermName) to the
TypeName of the member of the import's prefix.

retronym committed:
Show constructor def strings TAB,TAB, as we do for methods

retronym committed:
Merge pull request #7828 from retronym/topic/picklewriter

Avoid trailing zero bytes in .sig files written in PipelineMain

diesalbla committed:
Remove use of List.map in the "varianceInTypes" function.

The function `varianceInTypes`, used to calculate the variance of a
certain type parameter inside a type expression, used several calls
to `List.map`, which creates a new list that was immediately folded
into a single Variance result.

Using fold fusion, we remove the intermediate mapped lists.
We replace the several uses of `List.map` and the `fold` method from
the `Variance` companion objects by tail-recursive loops.
- By putting the `map` functiona and the fold together, it is not
  allocating any extra List.
- We add an "Extractor" trait in the Variance object, for functions
  that alwas return `Variance`. Because the return type is not
  polymorphic, unlike the `Function1`, this saves boxing/unboxing.
- We add a couple of polymorphic functions, `foldExtract` and `foldExtract2`,
  to get the variance intersection of a list, given the extractor
  for the given type.

diesalbla committed:
copyMethodCompleter: Replace map2 with foreach2

Inside the `copyMethodCompleter` we were using a call to the `map2`
method, that was generating a list. However, the result of that call
was not being used at all. Only the side-effects are relevant.
Thus, we can replace the `map2` with a `foreach2`.

diesalbla committed:
Avoid using List.zip in the SpecializeTypes unification

The code in the unify section of SpecializeTypes had a call to
`List.zip`, which was followed by a process to foldLeft that list of
pairs (by unifying the elements of each pair).
This is allocating a linear number of Cons `::` and Tuple2 objects.

We introduce in the Collections class a new method, foldLeft2,
to fold over two lists without extra allocations.
We use it to teplace the call to `zip`.

diesalbla committed:
Typers: by-pass the unzipped list of pairs.

The modified code in the Typers file was creating a list of pairs,
by using a `map2` function, which was immediately _unzipped_.
Thus, the first list was a linear number of unneeded allocations.

To avoid that intermediate list, we change the algorithm to directly
generate the two lists on a single pass. To do this:

- For the args1 list of trees, since the function can often return
  the input tree, unmodified, we use the main `mapConserve` function.
- For the args list main list, , we use mutable ListBuffer, and
  within the `mapConserve` function we have instructions to add them.

Co-authored-by: Jason Zaugg <jzaugg@gmail.com>

diesalbla committed:
Merge pull request #7837 from diesalbla/copy_method_completer_replace_map2_with_foreach2

copyMethodCompleter: Replace map2 with foreach2

retronym committed:
Reduce SAM lambda allocation within varianceInTypes

retronym committed:
Make varianceInType allocation free with a reusable instance

diesalbla committed:
Merge pull request #7838 from diesalbla/specialize_types_avoid_zip

Avoid using List.zip in the SpecializeTypes unification

retronym committed:
Merge pull request #7835 from diesalbla/variances_remove_listmap

Remove use of List.map in the "varianceInTypes" function.

retronym committed:
Merge pull request #7827 from diesalbla/uncurry_avoid_flatten

Uncurry: avoid using List.flatten

retronym committed:
Merge pull request #7809 from diesalbla/avoid_list_creation_

Avoid the call to BaseTypeSeq.toList in type Unification.

retronym committed:
Merge pull request #7820 from diesalbla/dedup_no_lists_backport

Namers - no duplicates: backport changes to avoid allocations [nomerge]

retronym committed:
Merge pull request #7826 from diesalbla/types_normalize_no_flatmap

Avoid using List flatMap within normalizeImpl

diesalbla committed:
Merge pull request #7840 from diesalbla/typers_avoid_unzipping

Typers: by-pass the unzipped list of pairs.

diesalbla committed:
Replace the uses of map for substInfo

The substInfo method from the `Symbol` class is a side-effectful
mutation: it modifies the internal data of the object that receives
it. As such, using a `map` to perform an operation on all elements
only creates an extra list that is not needed.

Thus, we replace the `map` with a `foreach`.

diesalbla committed:
Optimisation: replace mapParamss with a foreachParamss

We complement the `mapParamss` function of the Symbols cake slice with
a foreachParamss method, that performs a side-effectful action.
We replace several uses of mapParamss with the `foreachParamss`, to
avoid the extra allocations.

We also do other optimisations, such as merging a map followed by a
foldLeft into the fold, and a map followed by a foreach into the
foreach (which is another fold).

diesalbla committed:
Superaccessors: merge several List method.

The modified code was performing a `map`, followed by a `find`,
followed by a `foreach`. The first `map` was allocating an
intermediate list that was immediately consumed.
We use fusion to join those three operations in a single
foreach, to traverse without extra list allocations.

retronym committed:
Merge pull request #7852 from diesalbla/In_map_substInfo_use_foreach

Replace map with foreach, when using substInfo

retronym committed:
Merge pull request #7855 from diesalbla/merge_find_foreach_map

Superaccessors: merge several List method.

retronym committed:
Merge pull request #7854 from diesalbla/foreachParamss

Optimisation: replace mapParamss with a foreachParamss

diesalbla committed:
Definitions - SymbolSet class: deforest middle list.

The code to compute the `commonOwner` was using a List.map, that
allocates a new list, followed by a `distinct` operation, that may
generate a smaller one, followed by a singleton-list check.

We can achieve the same using a comparison between the head and all of
the elements in the tail, at no extra List allocation cost.

diesalbla committed:
Optimisations: avoid List allocations in the RefChecks.

We add some small optimisations to the code in the RefChecks
- We replace the `map length` calls with the utility methods in the
  Collections traits.
- We replace a combined use of flatten, map, zip, and filter with
  the use of an iterator and a special iterator function.
- We add to the Collections utils a function to create a special
  iterator, which combines zip, filter, and collect functions.

ennru committed:
[backport] Prefer HTTPS over HTTP for pom.xml data

diesalbla committed:
Remove List Allocations in the Checkable

Within the Checkable file, we add some changes to avoid some List
allocations.

- `propagateKnownTypes`: we merge the filter of base classes into the
  foreach block, as an if condition
- `typeArgsInTopLevel`: we replace the recursive List creation
  and flatMapping with a ListBuffer and a recursive loop,
  we merge a List `map` into a `foreach`, and we replace the
  `filterNot` by a guard condition when adding into the buffer.
- `allChildrenAreIrreconcilable`: we do not need to transform `toList`
  just to iterate (which forall does). However, we store the result of
  the `sealedChildren` since that can be expensive to compute.
- In the `isNeverSubArg`, we merge the `map` of variance into the
  `exists3` fold. This may also have the side benefit of avoiding
  allocations of the `Variance` value class.

retronym committed:
Merge pull request #7858 from diesalbla/Definitions_replace_map_distinct_check_singleton_with_head_tail_forall

Definitions - SymbolSet: avoid middle map list

retronym committed:
Merge pull request #7868 from diesalbla/Checkable_avoid_allocations

Remove List Allocations in the Checkable

retronym committed:
Merge pull request #7865 from diesalbla/RefChecks_avoid_flatten_maps_filter

Optimisations: avoid List allocations in the RefChecks.

lrytz committed:
Merge pull request #7872 from ennru/prefer-https-2.12

Backport of "Prefer HTTPS over HTTP for pom.xml data"

diesalbla committed:
Replace "zipped" by combinators without list allocations.

The `zipped` function performs a linear number of list allocations.
In some places, it was used to feed the result into a `map` or a
`foreach`. This commit replaces those usages with the `map2`,
or `foreach2` functions from the collections utility.

adriaanm commented on src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:
```diff
@@ -257,8 +257,10 @@ trait ScalaSettings extends AbsScalaSettings
   val YcacheMacroClassLoader   = CachePolicy.setting("macro", "macros")
-
   val YmacroClasspath = PathSetting       ("-Ymacro-classpath", "The classpath used to reflectively load macro implementations, default is the compilation classpath.", "")
+  val YpartialUnification = BooleanSetting ("-Ypartial-unification", "Enable partial unification in type constructor inference")
```
hmm, this shouldn't be revived

SethTisue committed:
Merge pull request #7833 from retronym/topic/completion-new

Improve REPL completion for class names and constructors

retronym committed:
Recycle the hashset used for implicit shadowing

hrhino committed:
Don't give unhelpful unemptiness advice.

As we've recently learned, sometimes `!isEmpty` is simply better than
`nonEmpty`, performance-wise. Thus, let IntelliJ not bring the yellow
highlight of shame down upon us for using it.

hrhino committed:
mailmap me!

retronym committed:
Merge pull request #7881 from diesalbla/zipped_map_convert_to_map2

Replace "zipped" by functions without list allocations.

retronym committed:
Refactor adjustTypeArgs, giving its result record a name

retronym committed:
Merge pull request #7898 from hrhino/topic/not-unempty

Don't give unhelpful unemptiness advice.

retronym committed:
Merge pull request #7890 from retronym/topic/implicit-shadower-reusable

Recycle the hashset used for implicit shadowing

retronym commented:
I'm looking at this PR now. Will submit an updated version tomorrow that resolves new conflicts and corrects the merge error.

retronym committed:
Avoid redundant computation of undetParams / refacator to avoid confusing similarly named idents

retronym committed:
Merge pull request #7900 from retronym/topic/adjusted-args

Refactor adjustTypeArgs, giving its result record a name

retronym committed:
Merge remote-tracking branch 'origin/2.13.x' into merge/2.12.x-to-2.12.x-20180305

retronym committed:
Merge commit '9862ab7a15' into merge/2.12.x-to-2.12.x-20180305

retronym committed:
Merge commit 'efb9faf7f6' into merge/2.12.x-to-2.12.x-20180305

retronym committed:
Merge commit 'f6930376f5' into merge/2.12.x-to-2.12.x-20180305

retronym committed:
Merge commit '6f4ee4a51d' into merge/2.12.x-to-2.12.x-20180305

retronym committed:
Merge commit '4acff24d35' into merge/2.12.x-to-2.12.x-20180305

retronym committed:
Merge commit '6990ee369a' into merge/2.12.x-to-2.12.x-20180305

retronym commented:
- New conflicts with 2.13.x resolved
 - `-Ypartial-unificaction` merge errror fixes
 - New commits from 2.12.x included in this merge

retronym committed:
Fix merge error from 10da62d

retronym committed:
Fix merge error in 3014cac

retronym commented:
@adriaanm This is ready now.

adriaanm commented:
I still see 

```diff
diff --git c/src/compiler/scala/tools/nsc/settings/ScalaSettings.scala w/src/compiler/scala/tools/nsc/settings/ScalaSettings.s
cala
index fa021cc..662dde6 100644
--- c/src/compiler/scala/tools/nsc/settings/ScalaSettings.scala
+++ w/src/compiler/scala/tools/nsc/settings/ScalaSettings.scala
@@ -257,8 +257,10 @@ trait ScalaSettings extends AbsScalaSettings
   val YdisableFlatCpCaching  = BooleanSetting    ("-YdisableFlatCpCaching", "Do not cache flat classpath representation of cl
asspath elements from jars across compiler instances.")
   val YcachePluginClassLoader  = CachePolicy.setting("plugin", "compiler plugins")
   val YcacheMacroClassLoader   = CachePolicy.setting("macro", "macros")
-
   val YmacroClasspath = PathSetting       ("-Ymacro-classpath", "The classpath used to reflectively load macro implementation
s, default is the compilation classpath.", "")
+  val YpartialUnification = BooleanSetting ("-Ypartial-unification", "Enable partial unification in type constructor inference")
+  val Yvirtpatmat     = BooleanSetting    ("-Yvirtpatmat", "Enable pattern matcher virtualization")
```

adriaanm commented:
I'll look into the curious tenacity of those -Y flags this morning.

retronym committed:
Really remove pesky -Y flags that have crept through merging/rebasing

retronym commented:
Egads.  I've hit them with my heftiest sledgehammer. `-Youtline` is gone too, it was for an experimental part of #7712 that I've deferred.

retronym committed:
Merge remote-tracking branch 'origin/2.13.x' into merge/2.12.x-to-2.12.x-20180305
